### PR TITLE
feat(appmesh): change Virtual Node service discovery to a union-like class

### PR DESCRIPTION
**Pull request by dfezzie**
Addresses part of the conversation on https://github.com/aws/aws-cdk/issues/9490

BREAKING CHANGE: the properties `dnsHostName` and `awsCloudMap` of `VirtualNodeProps` have been replaced with the property `serviceDiscovery`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/appmesh.ts [hunk 1]
```diff
@@ -279,7 +279,9 @@ export class AppMeshExtension extends ServiceExtension {
     this.virtualNode = new appmesh.VirtualNode(this.scope, `${this.parentService.id}-virtual-node`, {
       mesh: this.mesh,
       virtualNodeName: this.parentService.id,
-      cloudMapService: service.cloudMapService,
+      serviceDiscovery: service.cloudMapService? appmesh.ServiceDiscovery.cloudmap({
+        service: service.cloudMapService,
+      }): undefined,
```
Comment (skinny85) [inline packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/appmesh.ts:284]: Weird formatting... how about

```suggestion
      serviceDiscovery: service.cloudMapService
        ? appmesh.ServiceDiscovery.cloudmap({
          service: service.cloudMapService,
        })
       : undefined,
```

?

File: packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   * Returns DNS based service discovery
+   */
+  public static dns(hostname: string): ServiceDiscovery {
+    return new DnsServiceDiscovery(hostname);
+  }
+
+  /**
+   * Returns Cloud Map based service discovery
+   */
+  public static cloudmap(options: CloudMapServiceDiscoveryOptions): ServiceDiscovery {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts:59]: Should this be called `cloudMap` instead? That seems to be the casing we usually use for this service, and it's also called "Cloud Map", not "Cloudmap".

**Review by skinny85 — COMMENTED** [review]
Looks great @dfezzie ! One question, related to naming.

File: packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts (hunk 2 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts:59]: yep! Good catch. I'll make that change

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 3]
```diff
@@ -146,7 +146,9 @@ const namespace = new servicediscovery.PrivateDnsNamespace(this, 'test-namespace
 const service = namespace.createService('Svc');
 
 const node = mesh.addVirtualNode('virtual-node', {
-  cloudMapService: service,
+  serviceDiscovery: appmesh.ServiceDiscovery.cloudmap({
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:149]:
```suggestion
  serviceDiscovery: appmesh.ServiceDiscovery.cloudMap({
```

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 4]
```diff
@@ -168,7 +170,9 @@ Create a `VirtualNode` with the constructor and add tags.
 ```ts
 const node = new VirtualNode(this, 'node', {
   mesh,
-  cloudMapService: service,
+  serviceDiscovery: appmesh.ServiceDiscovery.cloudmap({
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:173]:
```suggestion
  serviceDiscovery: appmesh.ServiceDiscovery.cloudMap({
```

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 5]
```diff
@@ -198,7 +202,9 @@ Create a `VirtualNode` with the constructor and add backend virtual service.
...
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:205]:
```suggestion
  serviceDiscovery: appmesh.ServiceDiscovery.cloudmap({
```

**Review by skinny85 — COMMENTED** [review]
Looks good! Just minor changes to the ReadMe (previous name snuck by).

**Review by skinny85 — APPROVED** [review]
👍

**Comment by dfezzie** [comment]
Build failed, looks unrelated to these changes

**Pull request merged**